cmd/integrations/nagios/nagios_enqueue_test.go

```diff
@@ -0,0 +1,257 @@
+				notificationType: "RECOVERY",
+				sourceType:       "service",
+				customFields: map[string]string{
+					"HOSTNAME": "computer.network",
+				},
+			},
+			expectedError: errors.New("the SERVICEDESC field must be set for source-type \"service\" using the -f flag"),
+		},
+		{
+			name: "serviceStateNotSetServiceCustomDetails",
+			inputs: nagiosEnqueueInput{
+				routingKey:       "abc",
+				notificationType: "RECOVERY",
+				sourceType:       "service",
+				customFields: map[string]string{
+					"HOSTNAME":    "computer.network",
+					"SERVICEDESC": "a service",
+				},
+			},
+			expectedError: errors.New("the SERVICESTATE field must be set for source-type \"service\" using the -f flag"),
+		},
+		{
+			name: "hostnameNotSetHostCustomDetails",
+			inputs: nagiosEnqueueInput{
+				routingKey:       "abc",
+				notificationType: "RECOVERY",
+				sourceType:       "host",
+			},
+			expectedError: errors.New("the HOSTNAME field must be set for source-type \"host\" using the -f flag"),
+		},
+		{
+			name: "hoststateNotSetHostCustomDetails",
+			inputs: nagiosEnqueueInput{
+				routingKey:       "abc",
+				notificationType: "RECOVERY",
+				sourceType:       "host",
+				customFields: map[string]string{
+					"HOSTNAME": "computer.network",
+				},
+			},
+			expectedError: errors.New("the HOSTSTATE field must be set for source-type \"host\" using the -f flag"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			realConfig := cmdutil.NewConfig()
+
+			cmd := NewNagiosEnqueueCmd(realConfig)
+			cmd.SetArgs(buildCmdArgs(tt.inputs))
+
+			_, err := cmd.ExecuteC()
+
+			assert.Error(t, err)
+			assert.Equal(t, tt.expectedError, err)
+		})
+	}
+}
+
+func TestNagiosEnqueue_validInputs(t *testing.T) {
+	tests := []struct {
+		name      string
+		cmdInputs nagiosEnqueueInput
+	}{
+		{
+			name: "validSourceHostInput",
+			cmdInputs: nagiosEnqueueInput{
+				routingKey:       "xyz",
+				notificationType: "PROBLEM",
+				sourceType:       "host",
+				customFields: map[string]string{
+					"HOSTNAME":  "computer.network",
+					"HOSTSTATE": "down",
+				},
+			},
+		},
+		{
+			name: "validSourceServiceInput",
+			cmdInputs: nagiosEnqueueInput{
+				routingKey:       "xyz",
+				notificationType: "PROBLEM",
+				sourceType:       "service",
+				customFields: map[string]string{
+					"HOSTNAME":     "computer.network",
+					"SERVICESTATE": "down",
+					"SERVICEDESC":  "serviceA",
+				},
+			},
+		},
+		{
+			name: "userProvidedDedupKey",
+			cmdInputs: nagiosEnqueueInput{
+				routingKey:       "xyz",
+				notificationType: "PROBLEM",
+				sourceType:       "service",
+				dedupKey:         "somededupkey",
+				customFields: map[string]string{
+					"HOSTNAME":     "computer.network",
+					"SERVICESTATE": "down",
+					"SERVICEDESC":  "serviceA",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			viper.SetDefault("address", cmdutil.GetDefaults().Address)
```

Comment:
I'm interested in why this is required here.  (I see that if we remove it, the tests don't work but i'd like to understand why.)

As I am looking at this, i'm wondering if we also might want to spend a little time refactoring `cmdutil.Defaults` and `cmdutil.Config`

Reply:
Yeah for sure I'm also confused why this is necessary here, and not in the `cmd` tests. I did spend an hour or so trying to figure this out, but I'll put some more time into this today to see if I can get a definitive answer because this obviously isn't ideal.

Reply:
Good news! I figured why `address` was not getting set in the `nagios` package tests. the `cmd` package has an `init` function that gets run whenever the package is used to in `cmd` tests the viper stuff gets initialized. I pushed a few commits trying some things but this is the final solution I ended on. Lmk what you think: 33cdc38

Reply:
The solution you've put in place here is better.

As we discussed, we may want to do some further refactoring in the future to pull `viper` and its side-effects further to the edges of the application and have the inner layers access config through a config object.